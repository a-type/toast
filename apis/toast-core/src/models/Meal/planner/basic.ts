import { Schedule, Meal } from 'models';
import { MealActionType } from 'models/Meal/Meal';

export default (schedule: Schedule, meals: Meal[]): Meal[] => {
  meals.forEach(meal => {
    const ScheduleMeal = schedule.getMeal(meal.dayIndex, meal.mealIndex);
    const mealType = {
      LONG: 'FANCY',
      MEDIUM: 'NORMAL',
      SHORT: 'QUICK',
    }[ScheduleMeal.availability];
    if (mealType) {
      const cookActionData = {
        type: MealActionType.Cook,
        mealType,
        servings: schedule.defaultServings,
      };

      const cookAction = meal.addAction(cookActionData);

      const eatAction = {
        type: MealActionType.Eat,
        cookActionId: cookAction.id,
        leftovers: false,
      };

      meal.addAction(eatAction);
    } else {
      meal.addAction({
        type:
          ScheduleMeal.availability === 'EAT_OUT'
            ? MealActionType.EatOut
            : MealActionType.Skip,
      });
    }
  });

  return meals;
};
