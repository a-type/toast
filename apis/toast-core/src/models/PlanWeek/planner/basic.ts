import { Schedule, PlanWeek } from 'models';
import {
  PlanActionCook,
  PlanActionType,
  PlanActionEat,
  PlanAction,
} from '../types';
import { initializeWeek } from './utils';
import { id } from 'tools';

export default (schedule: Schedule, weekIndex: number): PlanWeek => {
  const week = initializeWeek(schedule, weekIndex, () => true);

  week.meals.forEach(meal => {
    const ScheduleMeal = schedule.getMeal(meal.dayIndex, meal.mealIndex);
    const mealType = {
      LONG: 'FANCY',
      MEDIUM: 'NORMAL',
      SHORT: 'QUICK',
    }[ScheduleMeal.availability];
    if (mealType) {
      const cookAction: PlanActionCook = {
        id: id('action'),
        type: PlanActionType.Cook,
        mealType,
        weekIndex,
        dayIndex: meal.dayIndex,
        mealIndex: meal.mealIndex,
        servings: schedule.defaultServings,
      };
      const eatAction: PlanActionEat = {
        id: id('action'),
        type: PlanActionType.Eat,
        weekIndex,
        dayIndex: meal.dayIndex,
        mealIndex: meal.mealIndex,
        cookActionId: cookAction.id,
        leftovers: false,
      };

      meal.actions.push(cookAction, eatAction);
    } else {
      meal.actions.push({
        id: id('action'),
        weekIndex,
        dayIndex: meal.dayIndex,
        mealIndex: meal.mealIndex,
        type:
          ScheduleMeal.availability === 'EAT_OUT'
            ? PlanActionType.EatOut
            : PlanActionType.Skip,
      } as PlanAction);
    }
  });

  return new PlanWeek(week);
};
