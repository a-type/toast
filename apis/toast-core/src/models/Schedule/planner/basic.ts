import { Schedule, Meal } from 'models';
import { MealActionType } from 'models/Meal/Meal';

export default (schedule: Schedule, meals: Meal[]): Meal[] => {
  meals.forEach(meal => {
    const scheduleMeal = schedule.getScheduleMeal(
      meal.dayIndex,
      meal.mealIndex,
    );

    if (!scheduleMeal) {
      return;
    }

    const recipeType = {
      LONG: 'FANCY',
      MEDIUM: 'NORMAL',
      SHORT: 'QUICK',
    }[scheduleMeal.availability];
    if (recipeType) {
      const cookActionData = {
        type: MealActionType.Cook,
        recipeType,
        servings: schedule.defaultServings,
      };

      const cookAction = meal.addAction(cookActionData);

      const eatAction = {
        type: MealActionType.Eat,
        cookActionId: cookAction.id,
        leftovers: false,
      };

      meal.addAction(eatAction);
    } else {
      meal.addAction({
        type:
          scheduleMeal.availability === 'EAT_OUT'
            ? MealActionType.EatOut
            : MealActionType.Skip,
      });
    }
  });

  return meals;
};
