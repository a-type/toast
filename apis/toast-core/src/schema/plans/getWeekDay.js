import { addDays, addWeeks } from 'date-fns';

export default ({ weekIndex, startDay, dayOffset }) => {
  const date = new Date(startDay.valueOf());
  return addDays(addWeeks(date, weekIndex), dayOffset);
};
