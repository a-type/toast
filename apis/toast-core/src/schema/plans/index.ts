import { gql } from 'apollo-server-express';
import { Context } from 'context';
import { path, mergeDeepRight } from 'ramda';
import { id } from 'tools';

import * as meals from './meals';
import * as schedules from './schedules';

export const typeDefs = () => [
  gql`
    type Plan {
      id: ID!
    }

    extend type Query {
      plan: Plan
    }
  `,
  meals.typeDefs,
  schedules.typeDefs,
];

export const resolvers = [
  {
    Query: {
      plan: async (_parent, args, ctx: Context) => {
        const group = await ctx.graph.groups.getMine();
        let planId = path<string>(['planId'], group);

        if (!planId) {
          planId = id('plan');
          await ctx.graph.groups.mergeMine({ planId });
        }

        const plan = await ctx.firestore.plans.get(planId);

        return plan;
      },
    },
  },
  meals.resolvers,
  schedules.resolvers,
].reduce(mergeDeepRight, {});
