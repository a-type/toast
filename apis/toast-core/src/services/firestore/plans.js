const COLLECTION = 'plans';
const START_WEEK_DAY = new Date(2018, 10, 7);

const defaulted = plan => ({
  servingsPerMeal: 0,
  days: [],
  ...plan,
});

export default class Plans {
  static START_WEEK_DAY = START_WEEK_DAY;
  START_WEEK_DAY = START_WEEK_DAY;

  constructor(firestore) {
    this.firestore = firestore;
  }

  /**
   * For user's default plan, use user id
   */
  get = async planId => {
    const document = this.firestore.doc(`${COLLECTION}/${planId}`);
    const docRef = await document.get();
    if (!docRef.exists) {
      return null;
    }

    return defaulted(docRef.data());
  };

  merge = async (planId, planData) => {
    const document = this.firestore.doc(`${COLLECTION}/${planId}`);
    await document.set(planData, { merge: true });
    const docRef = await document.get();
    return defaulted(docRef.data());
  };

  set = async (planId, plan) => {
    const document = this.firestore.doc(`${COLLECTION}/${planId}`);
    await document.set(plan);
    return defaulted(plan);
  };

  /**
   * Weeks - copies of the base plan, applied per-week
   */
  getWeek = async (planId, week) => {
    const document = this.firestore.doc(`${COLLECTION}/${planId}/week_${week}`);
    const docRef = await document.get();

    if (!docRef.exists) {
      // copy the base plan instead
      return await this.get(planId);
    }

    return defaulted(docRef.data());
  };

  mergeWeek = async (planId, week, planData) => {
    const document = this.firestore.doc(`${COLLECTION}/${planId}/week_${week}`);
    await document.set(planData, { merge: true });
    const docRef = await document.get();
    return defaulted(docRef.data());
  };
}
