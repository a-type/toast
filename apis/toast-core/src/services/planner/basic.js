export default plan => {
  const simpleActionAssignment = (meal, mealDay, mealIndex) => {
    switch (meal.availability) {
      case 'LONG':
        return [
          {
            type: 'COOK',
            servings: plan.servingsPerMeal,
            mealType: 'FANCY',
          },
          { type: 'EAT', mealIndex, mealDay },
        ];
      case 'MEDIUM':
        return [
          {
            type: 'COOK',
            servings: plan.servingsPerMeal,
            mealType: 'NORMAL',
          },
          { type: 'EAT', mealIndex, mealDay },
        ];
      case 'SHORT':
        return [
          {
            type: 'COOK',
            servings: plan.servingsPerMeal,
            mealType: 'QUICK',
          },
          { type: 'EAT', mealIndex, mealDay },
        ];
      case 'NONE':
      case 'EAT_OUT':
        return [{ type: 'EAT_OUT' }];
    }
  };

  return {
    ...plan,
    days: plan.days.map((day, idx) => ({
      ...day,
      meals: day.meals.map((meal, mealIndex) => ({
        ...meal,
        actions: simpleActionAssignment(meal, idx, mealIndex),
      })),
    })),
  };
};
