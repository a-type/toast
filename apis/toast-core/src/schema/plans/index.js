import { gql } from 'apollo-server-express';
import { id } from 'tools';
import { path, pathOr, assocPath, compose, mergeDeepLeft } from 'ramda';
import { UserInputError } from 'errors';
import getWeekIndex from './getWeekIndex';
import getWeekDay from './getWeekDay';
import Plan from 'services/planner/Plan';

const emptyPlan = {
  days: new Array(7).fill(null).map(() => ({
    meals: new Array(3).fill(null).map(() => ({
      availability: 'SKIP',
      actions: [],
    })),
  })),
};

export const typeDefs = gql`
  enum PrepAvailability {
    SKIP
    EAT_OUT
    NONE
    SHORT
    MEDIUM
    LONG
  }

  enum PlanActionType {
    EAT_OUT
    COOK
    EAT
    READY_MADE
    SKIP
  }

  enum PlanMealType {
    QUICK
    BIG
    FANCY
    NORMAL
  }

  enum PlanStrategy {
    BASIC
    PREP
    BIG_PREP
  }

  interface PlanAction {
    id: ID!
    type: PlanActionType!
    dayIndex: Int
    mealIndex: Int
  }

  type PlanActionEatOut implements PlanAction {
    id: ID!
    type: PlanActionType!
    dayIndex: Int
    mealIndex: Int
    note: String
  }

  type PlanActionCook implements PlanAction {
    id: ID!
    type: PlanActionType!
    dayIndex: Int
    mealIndex: Int
    servings: Int!
    mealType: PlanMealType!
    recipe: Recipe
  }

  type PlanActionEat implements PlanAction {
    id: ID!
    type: PlanActionType!
    dayIndex: Int
    mealIndex: Int
    leftovers: Boolean!
    cookAction: PlanActionCook
  }

  type PlanActionReadyMade implements PlanAction {
    id: ID!
    type: PlanActionType!
    dayIndex: Int
    mealIndex: Int
    note: String
  }

  type PlanActionSkip implements PlanAction {
    id: ID!
    type: PlanActionType!
    dayIndex: Int
    mealIndex: Int
  }

  type PlanMeal {
    id: ID!
    availability: PrepAvailability!
    actions: [PlanAction!]!
  }

  type PlanDay {
    id: ID!
    date: Date
    meals: [PlanMeal!]!
  }

  type Plan {
    id: ID!
    defaultServings: Int!
    days: [PlanDay!]!
    groceryDay: Int!
    warnings: [String!]!

    """
    Access weekly plans using the week index, which is based
    on the fixed plan chronology system. To calculate a week index
    for any particular day, use the root planWeekIndex query
    """
    week(weekIndex: Int!): Plan

    """
    A date representing the first day (Sunday) of this Plan week,
    if (and only if) this Plan is a weekly plan, and not the root "blueprint" plan
    """
    startDate: Date
  }

  input PlanSetDetailsInput {
    defaultServings: Int
    groceryDay: Int
  }

  input PlanSetMealDetailsInput {
    availability: PrepAvailability
  }

  extend type Mutation {
    setPlanDetails(details: PlanSetDetailsInput!): Plan!
      @hasScope(scope: "update:plan")
    setPlanMealDetails(
      dayIndex: Int!
      mealIndex: Int!
      details: PlanSetMealDetailsInput!
    ): Plan! @hasScope(scope: "update:plan")
    setPlanStrategy(strategy: PlanStrategy): Plan!

    setPlanActionRecipe(
      weekIndex: Int!
      dayIndex: Int!
      mealIndex: Int!
      actionId: ID!
      recipeId: ID!
    ): PlanAction!
  }

  extend type Group {
    plan: Plan @authenticated
  }

  extend type Query {
    """
    A shortcut function to determine the week index of any particular day
    within the planning system
    """
    planWeekIndex(year: Int!, month: Int!, date: Int!): Int!
    """
    The start date of the entire plan system. You can reference this
    to determine the chronology of plan weeks in conjunction with
    planWeekIndex and the weekIndex field on Plan itself
    """
    planStartWeekDate: Date!
  }
`;

export const resolvers = {
  Query: {
    planWeekIndex: (_parent, { year, month, date }, ctx) =>
      getWeekIndex({
        year,
        month,
        date,
        startDay: ctx.firestore.plans.START_WEEK_DAY,
      }),

    planStartWeekDate: (_parent, _args, ctx) =>
      ctx.firestore.plans.START_WEEK_DAY,
  },

  Mutation: {
    setPlanDetails: async (_parent, { details }, ctx) => {
      const group = await ctx.graph.groups.getMine();
      let planId = path(['planId'], group);
      // create plan if it doesn't exist
      if (!planId) {
        planId = id('plan');
        await ctx.graph.groups.mergeMine({ planId });
      }

      const defaultPlan = new Plan({ id: planId });
      const plan = (await ctx.firestore.plans.get(planId)) || defaultPlan;
      plan.defaultServings = details.defaultServings;
      plan.groceryDay = details.groceryDay;
      const processed = ctx.planner.run(plan);
      return ctx.firestore.plans.set(planId, processed);
    },

    setPlanMealDetails: async (
      _parent,
      { dayIndex, mealIndex, details },
      ctx,
    ) => {
      const group = await ctx.graph.groups.getMine();

      if (!group || !group.planId) {
        throw new UserInputError("You haven't created a plan yet");
      }

      const plan = await ctx.firestore.plans.get(group.planId);

      plan.setMealDetails(dayIndex, mealIndex, details);
      const processed = ctx.planner.run(plan);

      await ctx.firestore.plans.set(group.planId, plan);
      return plan;
    },

    setPlanStrategy: async (_parent, { strategy }, ctx) => {
      const group = await ctx.graph.groups.getMine();

      if (!group || !group.planId) {
        throw new UserInputError("You haven't created a plan yet");
      }

      const plan = await ctx.firestore.plans.get(group.planId);
      plan.strategy = strategy;
      const processed = ctx.planner.run(plan);
      await ctx.firestore.plans.set(group.planId, processed);
      return processed;
    },

    setPlanActionRecipe: async (
      _parent,
      { weekIndex, dayIndex, mealIndex, actionId, recipeId },
      ctx,
    ) => {
      const group = await ctx.graph.groups.getMine();

      if (!group || !group.planId) {
        throw new UserInputError("You haven't created a plan yet");
      }

      const plan = await ctx.firestore.plans.getWeek(group.planId, weekIndex);
      const action = plan.setActionRecipe(
        dayIndex,
        mealIndex,
        actionId,
        recipeId,
      );
      await ctx.firestore.plans.mergeWeek(group.planId, weekIndex, plan);
      return action;
    },
  },

  Group: {
    plan: async (parent, args, ctx) => {
      const { planId } = parent;
      const plan = await ctx.firestore.plans.get(planId);
      ctx.plan = plan;
      return plan;
    },
  },

  Plan: {
    week: async (parent, args, ctx) => {
      if (parent.weekIndex !== null) {
        throw new UserInputError(
          "You can't access the week field on a Plan which represents a week already",
        );
      }
      const week = await ctx.firestore.plans.getWeek(parent.id, args.weekIndex);
      ctx.week = week;
      return week;
    },

    startDate: (parent, _args, ctx) => {
      if (!parent.id.includes('week_')) {
        return null;
      }

      return getWeekDay({
        weekIndex: parent.weekIndex,
        startDay: ctx.firestore.plans.START_WEEK_DAY,
      });
    },
  },

  PlanAction: {
    __resolveType: obj => {
      switch (obj.type) {
        case 'COOK':
          return 'PlanActionCook';
        case 'EAT':
          return 'PlanActionEat';
        case 'EAT_OUT':
          return 'PlanActionEatOut';
        case 'READY_MADE':
          return 'PlanActionReadyMade';
        case 'SKIP':
          return 'PlanActionSkip';
      }
    },
  },

  PlanActionEat: {
    cookAction: async (parent, args, ctx) => {
      let plan = ctx.week || ctx.plan;
      if (!plan) {
        const group = await ctx.graph.groups.getMine();

        if (!group || !group.planId) {
          throw new UserInputError("You haven't created a plan yet");
        }

        plan = await ctx.firestore.plans.get(group.planId);
      }

      const { cookActionId } = parent;

      return plan.getAction(cookActionId);
    },
  },

  PlanActionCook: {
    recipe: async (parent, args, ctx) => {
      if (!parent.recipeId) {
        return null;
      }

      return ctx.graph.recipes.get(parent.recipeId);
    },
  },

  PlanDay: {
    date: (parent, args, ctx) => {
      if (ctx.week) {
        const weekIndex = ctx.week.weekIndex;
        const { dayIndex } = parent;

        const date = getWeekDay({
          weekIndex,
          startDay: ctx.firestore.plans.START_WEEK_DAY,
          dayOffset: dayIndex,
        });
        return date;
      }
      return null;
    },
  },
};
