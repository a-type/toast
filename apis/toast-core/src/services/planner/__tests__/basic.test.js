import basic from '../basic';
import { variety } from './plans';

describe('basic meal plan', () => {
  test('converts availability to meal types', () => {
    expect(basic(variety)).toMatchInlineSnapshot(`
Object {
  "days": Array [
    Object {
      "breakfast": Object {
        "actions": Array [
          Object {
            "type": "EAT_OUT",
          },
        ],
        "availability": "NONE",
        "meal": "breakfast",
      },
      "dinner": Object {
        "actions": Array [
          Object {
            "mealType": "FANCY",
            "servings": 2,
            "type": "COOK",
          },
          Object {
            "mealId": "0.dinner",
            "type": "EAT",
          },
        ],
        "availability": "LONG",
        "meal": "dinner",
      },
      "lunch": Object {
        "actions": Array [
          Object {
            "mealType": "QUICK",
            "servings": 2,
            "type": "COOK",
          },
          Object {
            "mealId": "0.lunch",
            "type": "EAT",
          },
        ],
        "availability": "SHORT",
        "meal": "lunch",
      },
    },
    Object {
      "breakfast": Object {
        "actions": Array [
          Object {
            "type": "EAT_OUT",
          },
        ],
        "availability": "NONE",
        "meal": "breakfast",
      },
      "dinner": Object {
        "actions": Array [
          Object {
            "mealType": "NORMAL",
            "servings": 2,
            "type": "COOK",
          },
          Object {
            "mealId": "1.dinner",
            "type": "EAT",
          },
        ],
        "availability": "MEDIUM",
        "meal": "dinner",
      },
      "lunch": Object {
        "actions": Array [
          Object {
            "type": "EAT_OUT",
          },
        ],
        "availability": "EAT_OUT",
        "meal": "lunch",
      },
    },
    Object {
      "breakfast": Object {
        "actions": Array [
          Object {
            "mealType": "QUICK",
            "servings": 2,
            "type": "COOK",
          },
          Object {
            "mealId": "2.breakfast",
            "type": "EAT",
          },
        ],
        "availability": "SHORT",
        "meal": "breakfast",
      },
      "dinner": Object {
        "actions": Array [
          Object {
            "mealType": "QUICK",
            "servings": 2,
            "type": "COOK",
          },
          Object {
            "mealId": "2.dinner",
            "type": "EAT",
          },
        ],
        "availability": "SHORT",
        "meal": "dinner",
      },
      "lunch": Object {
        "actions": Array [
          Object {
            "mealType": "QUICK",
            "servings": 2,
            "type": "COOK",
          },
          Object {
            "mealId": "2.lunch",
            "type": "EAT",
          },
        ],
        "availability": "SHORT",
        "meal": "lunch",
      },
    },
    Object {
      "breakfast": Object {
        "actions": Array [
          Object {
            "type": "EAT_OUT",
          },
        ],
        "availability": "NONE",
        "meal": "breakfast",
      },
      "dinner": Object {
        "actions": Array [
          Object {
            "type": "EAT_OUT",
          },
        ],
        "availability": "EAT_OUT",
        "meal": "dinner",
      },
      "lunch": Object {
        "actions": Array [
          Object {
            "type": "EAT_OUT",
          },
        ],
        "availability": "NONE",
        "meal": "lunch",
      },
    },
    Object {
      "breakfast": Object {
        "actions": Array [
          Object {
            "type": "EAT_OUT",
          },
        ],
        "availability": "NONE",
        "meal": "breakfast",
      },
      "dinner": Object {
        "actions": Array [
          Object {
            "mealType": "NORMAL",
            "servings": 2,
            "type": "COOK",
          },
          Object {
            "mealId": "4.dinner",
            "type": "EAT",
          },
        ],
        "availability": "MEDIUM",
        "meal": "dinner",
      },
      "lunch": Object {
        "actions": Array [
          Object {
            "type": "EAT_OUT",
          },
        ],
        "availability": "NONE",
        "meal": "lunch",
      },
    },
    Object {
      "breakfast": Object {
        "actions": Array [
          Object {
            "mealType": "FANCY",
            "servings": 2,
            "type": "COOK",
          },
          Object {
            "mealId": "5.breakfast",
            "type": "EAT",
          },
        ],
        "availability": "LONG",
        "meal": "breakfast",
      },
      "dinner": Object {
        "actions": Array [
          Object {
            "mealType": "FANCY",
            "servings": 2,
            "type": "COOK",
          },
          Object {
            "mealId": "5.dinner",
            "type": "EAT",
          },
        ],
        "availability": "LONG",
        "meal": "dinner",
      },
      "lunch": Object {
        "actions": Array [
          Object {
            "mealType": "QUICK",
            "servings": 2,
            "type": "COOK",
          },
          Object {
            "mealId": "5.lunch",
            "type": "EAT",
          },
        ],
        "availability": "SHORT",
        "meal": "lunch",
      },
    },
    Object {
      "breakfast": Object {
        "actions": Array [
          Object {
            "type": "EAT_OUT",
          },
        ],
        "availability": "NONE",
        "meal": "breakfast",
      },
      "dinner": Object {
        "actions": Array [
          Object {
            "mealType": "FANCY",
            "servings": 2,
            "type": "COOK",
          },
          Object {
            "mealId": "6.dinner",
            "type": "EAT",
          },
        ],
        "availability": "LONG",
        "meal": "dinner",
      },
      "lunch": Object {
        "actions": Array [
          Object {
            "type": "EAT_OUT",
          },
        ],
        "availability": "NONE",
        "meal": "lunch",
      },
    },
  ],
  "servingsPerMeal": 2,
}
`);
  });
});
