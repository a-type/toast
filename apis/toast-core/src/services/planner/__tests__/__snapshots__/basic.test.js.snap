// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic meal plan converts availability to meal types 1`] = `
Object {
  "days": Array [
    Object {
      "dayIndex": 0,
      "id": "plan_MOCK_ID_0-day_0",
      "meals": Array [
        Object {
          "actions": Array [
            Object {
              "dayIndex": 0,
              "id": "action_MOCK_ID_1",
              "mealIndex": 0,
              "type": "SKIP",
            },
          ],
          "availability": "NONE",
          "dayIndex": 0,
          "id": "plan_MOCK_ID_0-day_0-meal_0",
          "mealIndex": 0,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 0,
              "id": "action_MOCK_ID_2",
              "mealIndex": 1,
              "mealType": "QUICK",
              "servings": 2,
              "type": "COOK",
            },
            Object {
              "cookActionId": "action_MOCK_ID_2",
              "dayIndex": 0,
              "id": "action_MOCK_ID_3",
              "leftovers": false,
              "mealIndex": 1,
              "type": "EAT",
            },
          ],
          "availability": "SHORT",
          "dayIndex": 0,
          "id": "plan_MOCK_ID_0-day_0-meal_1",
          "mealIndex": 1,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 0,
              "id": "action_MOCK_ID_4",
              "mealIndex": 2,
              "mealType": "FANCY",
              "servings": 2,
              "type": "COOK",
            },
            Object {
              "cookActionId": "action_MOCK_ID_4",
              "dayIndex": 0,
              "id": "action_MOCK_ID_5",
              "leftovers": false,
              "mealIndex": 2,
              "type": "EAT",
            },
          ],
          "availability": "LONG",
          "dayIndex": 0,
          "id": "plan_MOCK_ID_0-day_0-meal_2",
          "mealIndex": 2,
          "servings": 2,
        },
      ],
    },
    Object {
      "dayIndex": 1,
      "id": "plan_MOCK_ID_0-day_1",
      "meals": Array [
        Object {
          "actions": Array [
            Object {
              "dayIndex": 1,
              "id": "action_MOCK_ID_6",
              "mealIndex": 0,
              "type": "SKIP",
            },
          ],
          "availability": "NONE",
          "dayIndex": 1,
          "id": "plan_MOCK_ID_0-day_1-meal_0",
          "mealIndex": 0,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 1,
              "id": "action_MOCK_ID_7",
              "mealIndex": 1,
              "type": "SKIP",
            },
          ],
          "availability": "EAT_OUT",
          "dayIndex": 1,
          "id": "plan_MOCK_ID_0-day_1-meal_1",
          "mealIndex": 1,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 1,
              "id": "action_MOCK_ID_8",
              "mealIndex": 2,
              "mealType": "NORMAL",
              "servings": 2,
              "type": "COOK",
            },
            Object {
              "cookActionId": "action_MOCK_ID_8",
              "dayIndex": 1,
              "id": "action_MOCK_ID_9",
              "leftovers": false,
              "mealIndex": 2,
              "type": "EAT",
            },
          ],
          "availability": "MEDIUM",
          "dayIndex": 1,
          "id": "plan_MOCK_ID_0-day_1-meal_2",
          "mealIndex": 2,
          "servings": 2,
        },
      ],
    },
    Object {
      "dayIndex": 2,
      "id": "plan_MOCK_ID_0-day_2",
      "meals": Array [
        Object {
          "actions": Array [
            Object {
              "dayIndex": 2,
              "id": "action_MOCK_ID_10",
              "mealIndex": 0,
              "mealType": "QUICK",
              "servings": 2,
              "type": "COOK",
            },
            Object {
              "cookActionId": "action_MOCK_ID_10",
              "dayIndex": 2,
              "id": "action_MOCK_ID_11",
              "leftovers": false,
              "mealIndex": 0,
              "type": "EAT",
            },
          ],
          "availability": "SHORT",
          "dayIndex": 2,
          "id": "plan_MOCK_ID_0-day_2-meal_0",
          "mealIndex": 0,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 2,
              "id": "action_MOCK_ID_12",
              "mealIndex": 1,
              "mealType": "QUICK",
              "servings": 2,
              "type": "COOK",
            },
            Object {
              "cookActionId": "action_MOCK_ID_12",
              "dayIndex": 2,
              "id": "action_MOCK_ID_13",
              "leftovers": false,
              "mealIndex": 1,
              "type": "EAT",
            },
          ],
          "availability": "SHORT",
          "dayIndex": 2,
          "id": "plan_MOCK_ID_0-day_2-meal_1",
          "mealIndex": 1,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 2,
              "id": "action_MOCK_ID_14",
              "mealIndex": 2,
              "mealType": "QUICK",
              "servings": 2,
              "type": "COOK",
            },
            Object {
              "cookActionId": "action_MOCK_ID_14",
              "dayIndex": 2,
              "id": "action_MOCK_ID_15",
              "leftovers": false,
              "mealIndex": 2,
              "type": "EAT",
            },
          ],
          "availability": "SHORT",
          "dayIndex": 2,
          "id": "plan_MOCK_ID_0-day_2-meal_2",
          "mealIndex": 2,
          "servings": 2,
        },
      ],
    },
    Object {
      "dayIndex": 3,
      "id": "plan_MOCK_ID_0-day_3",
      "meals": Array [
        Object {
          "actions": Array [
            Object {
              "dayIndex": 3,
              "id": "action_MOCK_ID_16",
              "mealIndex": 0,
              "type": "SKIP",
            },
          ],
          "availability": "NONE",
          "dayIndex": 3,
          "id": "plan_MOCK_ID_0-day_3-meal_0",
          "mealIndex": 0,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 3,
              "id": "action_MOCK_ID_17",
              "mealIndex": 1,
              "type": "SKIP",
            },
          ],
          "availability": "NONE",
          "dayIndex": 3,
          "id": "plan_MOCK_ID_0-day_3-meal_1",
          "mealIndex": 1,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 3,
              "id": "action_MOCK_ID_18",
              "mealIndex": 2,
              "type": "SKIP",
            },
          ],
          "availability": "EAT_OUT",
          "dayIndex": 3,
          "id": "plan_MOCK_ID_0-day_3-meal_2",
          "mealIndex": 2,
          "servings": 2,
        },
      ],
    },
    Object {
      "dayIndex": 4,
      "id": "plan_MOCK_ID_0-day_4",
      "meals": Array [
        Object {
          "actions": Array [
            Object {
              "dayIndex": 4,
              "id": "action_MOCK_ID_19",
              "mealIndex": 0,
              "type": "SKIP",
            },
          ],
          "availability": "NONE",
          "dayIndex": 4,
          "id": "plan_MOCK_ID_0-day_4-meal_0",
          "mealIndex": 0,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 4,
              "id": "action_MOCK_ID_20",
              "mealIndex": 1,
              "type": "SKIP",
            },
          ],
          "availability": "NONE",
          "dayIndex": 4,
          "id": "plan_MOCK_ID_0-day_4-meal_1",
          "mealIndex": 1,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 4,
              "id": "action_MOCK_ID_21",
              "mealIndex": 2,
              "mealType": "NORMAL",
              "servings": 2,
              "type": "COOK",
            },
            Object {
              "cookActionId": "action_MOCK_ID_21",
              "dayIndex": 4,
              "id": "action_MOCK_ID_22",
              "leftovers": false,
              "mealIndex": 2,
              "type": "EAT",
            },
          ],
          "availability": "MEDIUM",
          "dayIndex": 4,
          "id": "plan_MOCK_ID_0-day_4-meal_2",
          "mealIndex": 2,
          "servings": 2,
        },
      ],
    },
    Object {
      "dayIndex": 5,
      "id": "plan_MOCK_ID_0-day_5",
      "meals": Array [
        Object {
          "actions": Array [
            Object {
              "dayIndex": 5,
              "id": "action_MOCK_ID_23",
              "mealIndex": 0,
              "mealType": "FANCY",
              "servings": 2,
              "type": "COOK",
            },
            Object {
              "cookActionId": "action_MOCK_ID_23",
              "dayIndex": 5,
              "id": "action_MOCK_ID_24",
              "leftovers": false,
              "mealIndex": 0,
              "type": "EAT",
            },
          ],
          "availability": "LONG",
          "dayIndex": 5,
          "id": "plan_MOCK_ID_0-day_5-meal_0",
          "mealIndex": 0,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 5,
              "id": "action_MOCK_ID_25",
              "mealIndex": 1,
              "mealType": "QUICK",
              "servings": 2,
              "type": "COOK",
            },
            Object {
              "cookActionId": "action_MOCK_ID_25",
              "dayIndex": 5,
              "id": "action_MOCK_ID_26",
              "leftovers": false,
              "mealIndex": 1,
              "type": "EAT",
            },
          ],
          "availability": "SHORT",
          "dayIndex": 5,
          "id": "plan_MOCK_ID_0-day_5-meal_1",
          "mealIndex": 1,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 5,
              "id": "action_MOCK_ID_27",
              "mealIndex": 2,
              "mealType": "FANCY",
              "servings": 2,
              "type": "COOK",
            },
            Object {
              "cookActionId": "action_MOCK_ID_27",
              "dayIndex": 5,
              "id": "action_MOCK_ID_28",
              "leftovers": false,
              "mealIndex": 2,
              "type": "EAT",
            },
          ],
          "availability": "LONG",
          "dayIndex": 5,
          "id": "plan_MOCK_ID_0-day_5-meal_2",
          "mealIndex": 2,
          "servings": 2,
        },
      ],
    },
    Object {
      "dayIndex": 6,
      "id": "plan_MOCK_ID_0-day_6",
      "meals": Array [
        Object {
          "actions": Array [
            Object {
              "dayIndex": 6,
              "id": "action_MOCK_ID_29",
              "mealIndex": 0,
              "type": "SKIP",
            },
          ],
          "availability": "NONE",
          "dayIndex": 6,
          "id": "plan_MOCK_ID_0-day_6-meal_0",
          "mealIndex": 0,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 6,
              "id": "action_MOCK_ID_30",
              "mealIndex": 1,
              "type": "SKIP",
            },
          ],
          "availability": "NONE",
          "dayIndex": 6,
          "id": "plan_MOCK_ID_0-day_6-meal_1",
          "mealIndex": 1,
          "servings": 2,
        },
        Object {
          "actions": Array [
            Object {
              "dayIndex": 6,
              "id": "action_MOCK_ID_31",
              "mealIndex": 2,
              "mealType": "FANCY",
              "servings": 2,
              "type": "COOK",
            },
            Object {
              "cookActionId": "action_MOCK_ID_31",
              "dayIndex": 6,
              "id": "action_MOCK_ID_32",
              "leftovers": false,
              "mealIndex": 2,
              "type": "EAT",
            },
          ],
          "availability": "LONG",
          "dayIndex": 6,
          "id": "plan_MOCK_ID_0-day_6-meal_2",
          "mealIndex": 2,
          "servings": 2,
        },
      ],
    },
  ],
  "defaultServings": 2,
  "groceryDay": 0,
  "id": "plan_MOCK_ID_0",
  "isBlueprint": false,
  "strategy": "BASIC",
  "warnings": Array [],
  "weekIndex": null,
}
`;
